### ISSUE TITLE
Custom patches should not need to be formatted like a full page

### ISSUE THREAD

**Issue by dmaahs2017** [dbrgn/tealdeer#181]
Quote from PR #142 discussion:
> Another issue for a new PR: Right now the patches need to be formatted like a full page, including a heading. Otherwise, the first line is interpreted as the heading, and skipped when printing (because the page and the patch are passed to separate tokenizer instances). However, I think that too can be fixed in a separate PR, I don't want to hold up this PR too long.

Pre-requisites:
* Issue #180

Acceptance Criteria: 
* Format of .patch files should be expected to not have a heading
* Update tests to reflect this change

**Comment by dbrgn** [comment]
Thanks for creating this issue, I almost forgot about it 🙂

@dmaahs2017 are you interested in doing this change?

I see two ways to solve this.

Right now the tokenizer expects a file to start with a heading. Maybe we should parametrize the tokenizer to tell it whether to expect a page or a patch.

(We could also just tokenize whatever we get thrown at us, including - for example - headings in the middle of the page. However, slightly stricter parsing is probably a good idea.)

Another approach would be to concatenate the input data _before_ tokenizing it. Right now we just "print" the page and the patch in sequence, which results in a new tokenization / formatting step for both. If we'd concatenate the source, we could parse it as if it were a single page.

@niklasmohrin maybe you also have some good input here.

**Comment by dmaahs2017** [comment]
I'm leaning towards concatenating the text together then running it through the tokenizer, because when `--markdown` flag is enabled the raw markdown is output as if they are a single page, so the concatenation method is consistent with that. 

This would also be simpler to implement, and require no re-working of the Tokenizer. The downside that I see is that we'll have to Read all the pages into 1 String before sending that to the tokenizer.

**Comment by niklasmohrin** [comment]
I am against collecting all the pages into a string up front. I think that the `Tokenizer` (which just transforms an iterator of lines into an iterator of tokens). I would really like to have this as a `struct Tokens` that impls `Iterator` but also only takes an `Iterator<Item=&str>` instead of a `BufRead`. This is hard (impossible) though, because the this iterator would have to yield string slices for a set lifetime `'a` in which all references are valid. But the lifetimes of these references are _designed_ to be disjoint in order to buffer the lines. So that won't happen. (I think it can't hurt to implement `Iterator` because it provides all of the helper methods for no additional cost)

There is `.chain` on readers that chains two readers of any kind. The resulting struct also implements `BufRead`, so `Tokenizer` wouldn't have to be changed.

I would still try to abstract the line-iterator idea away from the implementation of `Tokenizer`. I can imagine a trait `LineReader` (that could easily be implemented for all `BufRead`s) that only provides a method `.lines()`. This would move the IO code out of the tokenizer and _could_ make tests of the tokenizer easier by implementing `LineReader` for dummy types. It would work without the abstraction too though (I think :sweat_smile: ).

---

If we don't want to / have to keep state inbetween pages and patches, I like your idea of providing two methods that handle these two cases. If we make `Tokenizer` (which should then maybe be renamed) an iterator, we could just `Iterator::chain` the two iterators.

---

Let me know what you think!

**Comment by dmaahs2017** [comment]
I have a proposed solution. I drafted a spike in PR: #202. It's just a quick spike so if it's not the direction we want I won't feel bad to discard it.

Essentially I just created an intermediate type `SeqFileReader` which holds multiple `File` handles and implements `Read`, such that it reads the files in order as if they were 1 file. Then we can create a `BufReader` from the `SeqFileReader` just as if we were creating a `BufReader` from a single `File` type.

Issues with the spike solution as it is:
  - multiple file handles open at the same time, but at the moment the max would be 2, since `PageLookupResult` only returns max 2 paths for the time being.
  - Tokenizer may have to be adapted to detect TLDR Page version difference between the Page and the Patch file. (Since from the tokenizers perspective it is just reading 1 contiguous file)
  - No whitespace at the end of the page or beginning of the patch file means that 2 items will be placed right next to each other

Advantages:
- Simple implementation (KISS)
- cohesive with the current process

This seems like it could be promising but I don't want to put more work into it if it's not the desired direction. Would love comment on it.

TLDR; Abstract multiple Files (page and patch) into a single reader; the tokenizer doesn't need to know the difference. See #202

**Comment by dbrgn** [comment]
I agree, I don't see a big problem with opening multiple file handles. And not being able to open an existing patch file is quite unlikely.

I also agree that we should handle patches on the reader or iterator level, and not in the tokenizer.

Let's try to get #189 finished first, then we can discuss the reader again.

**Comment by dmaahs2017** [comment]
Ah very cool, didn't know about the chain method on the Read trait! Indeed we do not need SeqFileReader in that case :)

**Comment by niklasmohrin** [comment]
I would still be interested whether there is a (performance) improvement when chaining the files instead of chaining the iterators over each files lines. That would be something that a PR for this issue should investigate :)

If the two approaches perform the same (as I would expect), we can pick whichever one we like best

**Comment by dbrgn** [comment]
Hey @dmaahs2017, sorry for the long delay in reviewing. I like your approach and the proposal in #202, doing this on the I/O level is the proper solution.

@niklasmohrin is right that `chain` might be sufficient for our use case, so I'll try to come up with an implementation using `chain` and may re-use some parts of your code in #202.

**Comment by dmaahs2017** [comment]
> Hey @dmaahs2017, sorry for the long delay in reviewing. I like your approach and the proposal in #202, doing this on the I/O level is the proper solution.
> 
> @niklasmohrin is right that `chain` might be sufficient for our use case, so I'll try to come up with an implementation using `chain` and may re-use some parts of your code in #202.

@dbrgn Sounds good. Feel free to close #202 if/when it becomes redundant. I've been quite busy at work/school and don't have much time to contribute lately

**Comment by dbrgn** [comment]
PR here: https://github.com/dbrgn/tealdeer/pull/238

@dmaahs2017 I could mostly re-use your tests and added you as co-author of that commit.
